cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ApplicationLifecycleService.java

```diff
@@ -224,6 +229,12 @@ protected void shutDown() throws Exception {
     List<ApplicationDetail> result = new ArrayList<>();
     for (Map.Entry<ApplicationId, ApplicationSpecification> entry : appSpecs.entrySet()) {
       ApplicationDetail applicationDetail = ApplicationDetail.fromSpec(entry.getValue(), owners.get(entry.getKey()));
+      try {
+        capabilityReader.ensureApplicationEnabled(namespace.getNamespace(), applicationDetail.getName());
```

Comment:
Are we always hiding apps when capability is disable?

Reply:
Also, if the `appClass.getRequirements().getCapabilities()` already tells what capability an app needs, then why we need the `ensureApplicationEnabled` method?

Reply:
Changed to use capabilities from  appspec